File: frontend/src/types/quiz.ts
export interface Quiz {
    id: number;
    lesson_id: number | null;
    disaster_id: number | null;
    title: string;
    description: string | null;
    passing_score: number;
    time_limit_minutes: number | null;
    is_published: boolean;
    created_at: string;
}


export interface QuizOption {
    id: number;
    option_text: string;
    option_order: number;
}


export interface AdminQuizOption extends QuizOption {
    question_id: number;
    is_correct: boolean;
}


export interface QuizQuestion {
    id: number;
    question_text: string;
    question_order: number;
    points: number;
    options: QuizOption[];
}


export interface AdminQuestion extends QuizQuestion {
    quiz_id: number;
    options: AdminQuizOption[];
}


export interface StudentQuizDetail extends Quiz {
    questions: QuizQuestion[];
}


export interface AdminQuizDetail extends Quiz {
    questions: AdminQuestion[];
}


export interface QuizCreate {
    lesson_id?: number | null;
    disaster_id?: number | null;
    title: string;
    description?: string | null;
    passing_score: number;
    time_limit_minutes?: number | null;
}


export interface QuizUpdate {
    title?: string;
    description?: string | null;
    passing_score?: number;
    time_limit_minutes?: number | null;
}


export interface QuizOptionCreate {
    option_text: string;
    option_order: number;
    is_correct: boolean;
}


export interface QuestionCreate {
    question_text: string;
    question_order: number;
    points: number;
    options: QuizOptionCreate[];
}


export interface StartQuizResponse {
    attempt_id: number;
    quiz_id: number;
    started_at: string;
}


export interface SubmitAnswer {
    question_id: number;
    selected_option_id: number | null;
}


export interface SubmitQuizRequest {
    answers: SubmitAnswer[];
}


export interface QuizResult {
    attempt_id: number;
    quiz_id: number;
    score: number;
    total_points: number;
    percentage: number;
    passed: boolean;
    completed_at: string;
}

File: backend/app/schemas/quiz.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# =========================
# Quiz Create / Update
# =========================

class QuizCreate(BaseModel):
    lesson_id: int | None = None
    disaster_id: int | None = None
    title: str = Field(min_length=2, max_length=200)
    description: str | None = None
    passing_score: int = Field(default=60, ge=0, le=100)
    time_limit_minutes: int | None = Field(default=None, ge=1)


class QuizUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=200)
    description: str | None = None
    passing_score: int | None = Field(default=None, ge=0, le=100)
    time_limit_minutes: int | None = Field(default=None, ge=1)


class QuizResponse(BaseModel):
    id: int
    lesson_id: int | None
    disaster_id: int | None
    title: str
    description: str | None
    passing_score: int
    time_limit_minutes: int | None
    is_published: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =========================
# Create Question
# =========================

class QuizOptionCreate(BaseModel):
    option_text: str = Field(min_length=1, max_length=500)
    option_order: int = Field(ge=1)
    is_correct: bool = False


class QuestionCreate(BaseModel):
    question_text: str = Field(min_length=3)
    question_order: int = Field(ge=1)
    points: int = Field(default=1, ge=1)
    options: list[QuizOptionCreate] = Field(min_length=2)


# =========================
# Admin / Faculty Responses
# =========================

class AdminQuizOptionResponse(BaseModel):
    id: int
    question_id: int
    option_text: str
    option_order: int
    is_correct: bool

    model_config = ConfigDict(from_attributes=True)


class AdminQuestionResponse(BaseModel):
    id: int
    quiz_id: int
    question_text: str
    question_order: int
    points: int
    options: list[AdminQuizOptionResponse]

    model_config = ConfigDict(from_attributes=True)


class AdminQuizDetailResponse(QuizResponse):
    questions: list[AdminQuestionResponse] = []


# =========================
# Student Responses
# =========================

class StudentQuizOptionResponse(BaseModel):
    id: int
    option_text: str
    option_order: int

    model_config = ConfigDict(from_attributes=True)


class StudentQuestionResponse(BaseModel):
    id: int
    question_text: str
    question_order: int
    points: int
    options: list[StudentQuizOptionResponse]

    model_config = ConfigDict(from_attributes=True)


class StudentQuizDetailResponse(QuizResponse):
    questions: list[StudentQuestionResponse] = []


# =========================
# Attempts
# =========================

class StartQuizResponse(BaseModel):
    attempt_id: int
    quiz_id: int
    started_at: datetime


class SubmitAnswer(BaseModel):
    question_id: int
    selected_option_id: int | None = None


class SubmitQuizRequest(BaseModel):
    answers: list[SubmitAnswer]


class QuizResultResponse(BaseModel):
    attempt_id: int
    quiz_id: int
    score: int
    total_points: int
    percentage: float
    passed: bool
    completed_at: datetime